File: win-prog/ProgWinCS/Lines Curves and Area Fills/Spiral/Spiral.cs
//-------------------------------------
// Spiral.cs © 2001 by Charles Petzold
//-------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class Spiral: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new Spiral());
     }
     public Spiral()
     {
          Text = "Spiral";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          const int iNumRevs   = 20;
          int       iNumPoints = iNumRevs * 2 * (cx + cy);
          PointF[]  aptf       = new PointF[iNumPoints];
          float     fAngle, fScale;

          for (int i = 0; i < iNumPoints; i++)
          {
               fAngle = (float)(i * 2 * Math.PI /(iNumPoints / iNumRevs));
               fScale = 1 - (float)i / iNumPoints;

               aptf[i].X = (float)(cx / 2 * (1 + fScale * Math.Cos(fAngle)));
               aptf[i].Y = (float)(cy / 2 * (1 + fScale * Math.Sin(fAngle)));
          }
          grfx.DrawLines(new Pen(clr), aptf);
     }
}

File: win-prog/ProgWinVB/Lines Curves and Area Fills/Spiral/Spiral.vb
'---------------------------------------
' Spiral.vb (c) 2002 by Charles Petzold
'---------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class Spiral
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(New Spiral())
    End Sub

    Sub New()
        Text = "Spiral"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        Const iNumRevs As Integer = 20
        Dim iNumPoints As Integer = iNumRevs * 2 * (cx + cy)
        Dim aptf(iNumPoints) As PointF
        Dim rAngle, rScale As Double
        Dim i As Integer

        For i = 0 To iNumPoints
            rAngle = i * 2 * Math.PI / (iNumPoints / iNumRevs)
            rScale = 1 - i / iNumPoints
            aptf(i).X = CSng(cx / 2 * (1 + rScale * Math.Cos(rAngle)))
            aptf(i).Y = CSng(cy / 2 * (1 + rScale * Math.Sin(rAngle)))
        Next i

        grfx.DrawLines(New Pen(clr), aptf)
    End Sub
End Class
